azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/PartitionManager.java

```diff
@@ -95,339 +123,370 @@ void onPartitionCheckCompleteTestHook()
-                        	allLeases.put(possibleLease.getPartitionId(), possibleLease);
-                        	leasesOwnedByOthers.add(possibleLease);
-                    	}
-                    }
-                    else
-                    {
-                    	allLeases.put(possibleLease.getPartitionId(), possibleLease);
-                    	leasesOwnedByOthers.add(possibleLease);
-                    }
-            	}
-        		// Most exceptions will arrive packaged as ExecutionException because they occur during a short-lived thread
-        		// down in AzureStorageCheckpointLeastManager. However, AzureBlobLease.isExpired calls Storage directly and
-        		// therefore can throw a plain StorageException. Handling is the same for all: log, notify, and move on to the
-            	// next partition.
-            	catch (ExecutionException|StorageException e)
-            	{
-            		TRACE_LOGGER.warn(LoggingUtils.withHost(this.host.getHostName(), "Failure getting/acquiring/renewing lease, skipping"), e);
-            		Exception notifyWith = e;
-            		if ((e instanceof ExecutionException) && (e.getCause() != null) && (e.getCause() instanceof Exception))
-            		{
-            			notifyWith = (Exception)e.getCause();
-            		}
+    	// DO NOT check whether this.scanFuture is cancelled. The first execution of this method is scheduled
+    	// with 0 delay and can occur before this.scanFuture is set to the result of the schedule() call.
+
+        ArrayList<CompletableFuture<Lease>> allLeasesResults = new ArrayList<CompletableFuture<Lease>>();
+        // These are final so they can be used in the lambdas below.
+        final AtomicInteger ourLeasesCount = new AtomicInteger();
+        final ConcurrentHashMap<String, Lease> leasesOwnedByOthers = new ConcurrentHashMap<String, Lease>();
+        final BoolWrapper resultsAreComplete = new BoolWrapper(true);
+
+    	List<CompletableFuture<Lease>> gettingAllLeases = this.host.getLeaseManager().getAllLeases(this.partitionIds);
+
+        for (CompletableFuture<Lease> leaseFuture : gettingAllLeases)
+        {
+        	// Stage 0: get the lease
+        	CompletableFuture<Lease> oneResult = leaseFuture
+        	// Stage 1: is the lease expired?
+        	.thenComposeAsync((l) -> l.isExpired(), this.host.getExecutorService())
+        	// Stage 2: combine the two previous results so stage 3 can consume both
+        	.thenCombineAsync(leaseFuture, (exists, lease) ->
+        	{
+        		return new LeaseContext(exists, lease);
+        	}, this.host.getExecutorService())
+        	// Stage 3: if the lease is expired, acquire it and return success (true) or failure (false); always return false if not expired
+        	.thenComposeAsync((lc) ->
+        	{
+        		return lc.shouldDoNextStep ? this.host.getLeaseManager().acquireLease(lc.lease) : CompletableFuture.completedFuture(false);
```

Comment:
>return lc.shouldDoNextStep ? this.host.getLeaseManager().acquireLease(lc.lease) : CompletableFuture.completedFuture(false); [](start = 10, length = 123)

Why did we adapt acquire-then-steal pattern?
Why not "cautiously" acquire - keeping hostCount in perspective.